core/src/main/java/org/jdbi/v3/core/mapper/OptionalMapperFactory.java

```diff
@@ -38,28 +42,59 @@
  * </ul>
  */
 class OptionalMapperFactory implements ColumnMapperFactory {
-    private final Map<Class<?>, ColumnMapper<?>> mappers = new IdentityHashMap<>();
-
-    OptionalMapperFactory() {
-        mappers.put(OptionalInt.class, optionalMapper(ResultSet::getInt, OptionalInt::empty, OptionalInt::of));
-        mappers.put(OptionalLong.class, optionalMapper(ResultSet::getLong, OptionalLong::empty, OptionalLong::of));
-        mappers.put(OptionalDouble.class, optionalMapper(ResultSet::getDouble, OptionalDouble::empty, OptionalDouble::of));
-    }
+    private static final Map<Class<?>, OptionalStrategy> STRATEGIES =
+            Collections.unmodifiableMap(Arrays.stream(OptionalStrategy.values())
+                .collect(Collectors.toMap(
+                        OptionalStrategy::optClass,
+                        Function.identity(),
+                        (u, v) -> {
+                            throw new UnsupportedOperationException();
+                        },
+                        IdentityHashMap::new)));
 
     @Override
     public Optional<ColumnMapper<?>> build(Type type, ConfigRegistry config) {
-        Class<?> rawType = getErasedType(type);
+        return Optional.ofNullable(STRATEGIES.get(getErasedType(type)))
+                .map(strategy -> strategy.create(type, config));
+    }
+
+    private enum OptionalStrategy {
+        OBJ(Optional.class) {
+            @Override
+            ColumnMapper<?> create(Type type, ConfigRegistry config) {
+                final ColumnMapper<?> mapper = config.get(ColumnMappers.class).findFor(
+                        GenericTypes.findGenericParameter(type, Optional.class)
+                            .orElseThrow(() -> new NoSuchMapperException("No mapper for raw Optional type")))
```

Comment:
This also means that configuration errors (mapping `Optional<T>` when `T` isn't mappable) are checked eagerly.